rollout/service.go

```diff
@@ -277,13 +279,34 @@ func (c *rolloutContext) ensureSVCTargets(svcName string, rs *appsv1.ReplicaSet,
 	}
 	currSelector := svc.Spec.Selector[v1alpha1.DefaultRolloutUniqueLabelKey]
 	desiredSelector := rs.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
-	if currSelector != desiredSelector {
+	logCtx := c.log.WithField(logutil.ServiceKey, svc.Name)
+
+	if _, ok := svc.Annotations[v1alpha1.ManagedByRolloutsKey]; !ok && currSelector != desiredSelector {
+		// This if block will be entered only when adopting a service that already exists, because the current annotation
+		// will be empty at that point. When we are adopting a service, we want to make sure that the replicaset is fully
+		// available before we start routing traffic to it, so we do not overload it.
+		// See PR: https://github.com/argoproj/argo-rollouts/pull/1777
+
 		// ensure ReplicaSet is fully available, otherwise we will point the service to nothing or an underprovisioned ReplicaSet
 		if checkRsAvailability && !replicasetutil.IsReplicaSetAvailable(rs) {
-			logCtx := c.log.WithField(logutil.ServiceKey, svc.Name)
 			logCtx.Infof("delaying service switch from %s to %s: ReplicaSet not fully available", currSelector, desiredSelector)
 			return nil
 		}
+		logCtx.Infof("adopting service %s", svc.Name)
+		err = c.switchServiceSelector(svc, desiredSelector, c.rollout)
+		if err != nil {
+			return err
+		}
+	} else if currSelector != desiredSelector {
+		// This if block will be called only at the beginning and end of a rollout that is adopted, when we are at the end of a rollout
+		// we generally will have enough capacity to handle the traffic, so we do not need to check the full availability of the
+		// ReplicaSet. We do still want to make sure we have at least one pod available, so we do not point the service to nothing, but
+		// losing a pod or two should be tolerable to still switch service selectors.
+
+		if checkRsAvailability && !replicasetutil.IsReplicaSetPartiallyAvailable(rs) {
+			logCtx.Infof("delaying service switch from %s to %s: ReplicaSet has zero availability", currSelector, desiredSelector)
+			return nil
```

Comment:
When service switch is delayed, should it return err, so that it will not move on to the next step `reconcileTrafficRouting`?

Reply:
Yes that was actually one of the first things that was tried and also dose fix the issue is by introducing this error state however there are dependencies within rollouts that we still have to continue reconciling to keep progressing. There is quite a bit of background here https://github.com/argoproj/argo-rollouts/pull/2187 as well as here https://github.com/argoproj/argo-rollouts/pull/1777 where the intial return nil was introduced. It was introduced to prevent the adoption of a service causing downtime by making sure the service was fully available. The change in this PR keeps that behavior for the adoption case but switches to only needing some availability for the cases that we have already adopted the service.